File: franchise_erp/public/js/subcontracting_order.js
frappe.ui.form.on("Subcontracting Order", {
    refresh(frm) {
       
        frm.add_custom_button(
            __("Fabric Wastage Register"),
            function () {
                frappe.new_doc("Fabric Wastage Register", {
                    subcontracting_order: frm.doc.name,
                    company: frm.doc.company,
                    supplier: frm.doc.supplier
                });
            },
            __("Create")
        );
        
        if (frm.doc.docstatus !== 0) {
            return;
        }

        set_all_item_tax_templates(frm);

        if (
            frm.doc.purchase_order &&
            frm.doc.company &&
            !frm.doc.taxes_and_charges
        ) {
            set_taxes_and_charges(frm);
        } else {
            set_tax_rows_on_net_total(frm);
        }


    },

    purchase_order(frm) {
        if (frm.doc.purchase_order && frm.doc.company) {
            set_taxes_and_charges(frm);
        }
    },

    company(frm) {
        if (!frm.doc.company) {
            return;
        }

        set_all_item_tax_templates(frm);

        if (frm.doc.purchase_order) {
            set_taxes_and_charges(frm);
        }
    },

    taxes_and_charges(frm) {
        setTimeout(() => {
            set_tax_rows_on_net_total(frm);
        }, 500);
    },

    validate(frm) {
        set_tax_rows_on_net_total(frm);
    }
});


frappe.ui.form.on("Subcontracting Order Item", {
    item_code(frm, cdt, cdn) {
        set_item_tax_template(frm, cdt, cdn);
    },

    rate(frm, cdt, cdn) {
        set_item_tax_template(frm, cdt, cdn);
    }
});


function set_all_item_tax_templates(frm) {
    (frm.doc.items || []).forEach(row => {
        if (row.item_code) {
            set_item_tax_template(
                frm,
                row.doctype,
                row.name
            );
        }
    });
}


function set_item_tax_template(frm, cdt, cdn) {
    const row = locals[cdt][cdn];

    if (!row) {
        return;
    }

    if (!row.item_code || !frm.doc.company) {
        if (row.item_tax_template) {
            frappe.model.set_value(
                cdt,
                cdn,
                "item_tax_template",
                ""
            );
        }

        return;
    }

    frappe.call({
        method:
            "franchise_erp.custom.subcontracting_order.get_item_tax_template",

        args: {
            item_code: row.item_code,
            rate: row.rate || 0,
            company: frm.doc.company
        },

        callback(r) {
            const template = r.message || "";

            if (template !== row.item_tax_template) {
                frappe.model.set_value(
                    cdt,
                    cdn,
                    "item_tax_template",
                    template
                );
            }
        }
    });
}


function set_taxes_and_charges(frm) {
    if (!frm.doc.purchase_order || !frm.doc.company) {
        return;
    }

    frappe.call({
        method:
            "franchise_erp.custom.subcontracting_order.get_job_work_order_taxes_and_charges",

        args: {
            purchase_order: frm.doc.purchase_order,
            company: frm.doc.company
        },

        callback(r) {
            const template = r.message;

            if (!template) {
                return;
            }

            if (template !== frm.doc.taxes_and_charges) {
                frm.set_value("taxes_and_charges", template)
                    .then(() => {
                        setTimeout(() => {
                            set_tax_rows_on_net_total(frm);
                        }, 500);
                    });
            } else {
                set_tax_rows_on_net_total(frm);
            }
        }
    });
}


function set_tax_rows_on_net_total(frm) {
    if (!frm.doc.taxes || !frm.doc.taxes.length) {
        return;
    }

    let changed = false;

    frm.doc.taxes.forEach(row => {
        if (
            row.charge_type === "Actual" ||
            row.charge_type === "On Previous Row Total"
        ) {
            row.charge_type = "On Net Total";
            changed = true;
        }
    });

    if (changed) {
        frm.refresh_field("taxes");
    }
}

frappe.ui.form.on('Subcontracting Order', {
    refresh(frm) {

        // 🛑 1. Unsaved / New document guard
        if (frm.is_new()) return;

        // 🛑 2. Only Draft documents with Supplier
        if (frm.doc.docstatus !== 0 || !frm.doc.supplier) return;

        // 🧹 3. Prevent duplicate buttons
        frm.remove_custom_button(__('Outgoing Logistics'), __('Create'));
        
        // 🔍 4. Check if submitted Stock Entry exists
        frappe.db.get_value(
            "Stock Entry",
            {
                subcontracting_order: frm.doc.name,
                docstatus: 1
            },
            "name"
        ).then(se => {

            // ❌ No Stock Entry → no button
            if (!se.message) return;

            // 🔍 5. Check Supplier flag
            frappe.db.get_value(
                "Supplier",
                frm.doc.supplier,
                "custom_gate_out_applicable"
            ).then(r => {

                if (!r.message || !r.message.custom_gate_out_applicable) return;

                // ➕ 6. Add Create → Outgoing Logistics button
                frm.add_custom_button(
                    __('Outgoing Logistics'),
                    () => {
                        frappe.call({
                            method: "franchise_erp.custom.subcontracting_order.get_outgoing_logistics_data",
                            args: {
                                subcontracting_order: frm.doc.name
                            },
                            freeze: true,
                            callback(res) {
                                if (res.message) {
                                    frappe.new_doc("Outgoing Logistics", res.message);
                                }
                            }
                        });
                    },
                    __('Create')
                );
            });
        });
    }
});

File: franchise_erp/custom/subcontracting_order.py
import frappe

# @frappe.whitelist()
# def get_outgoing_logistics_data(subcontracting_order):
#     sc = frappe.get_doc("Subcontracting Order", subcontracting_order)

#     # Prevent duplicate
#     existing = frappe.db.exists(
#         "Outgoing Logistics",
#         {"document_no": sc.name}
#     )
#     if existing:
#         frappe.throw(f"Outgoing Logistics already exists: {existing}")

#     return {
#         "owner_site": sc.company,
#         "company_abbreviation": frappe.db.get_value("Company", sc.company, "abbr"),
#         "consignee_supplier": sc.supplier,
#         "transporter": sc.supplier,
#         "date": frappe.utils.today(),
#         "document_no": sc.name,
#         # "document_date": sc.transaction_date,
#         "quantity": sc.total_qty,
#         "unit": "Nos",
#         # "type": "S&D: Sales Invoice/Transfer In",
#         "type": "Job Order",
#         "mode": "Land",
#     }
import frappe
from frappe.utils import today

@frappe.whitelist()
def get_outgoing_logistics_data(subcontracting_order):

    if not subcontracting_order:
        frappe.throw("Subcontracting Order is required")

    # ✅ Safe doc fetch
    sc = frappe.get_doc("Subcontracting Order", subcontracting_order)

    return {
        "owner_site": sc.company,
        "company_abbreviation": frappe.db.get_value(
            "Company",
            sc.company,
            "abbr"
        ),
        "consignee_supplier": sc.supplier,
        "transporter": sc.supplier,
        "date": today(),
        "quantity": sc.total_qty,
        "unit": "Nos",
        "type": "Job Order",
        "mode": "Land",
        "references": [
            {
                "source_doctype": "Job Work Order",
                "source_name": sc.name
            }
        ]
    }





import frappe

@frappe.whitelist()
def get_subcontracting_order_city(subcontracting_order):
    so = frappe.get_doc("Subcontracting Order", subcontracting_order)

    # 1️⃣ Shipping Address (priority)
    if so.shipping_address:
        city = frappe.db.get_value(
            "Address",
            so.shipping_address,
            "custom_citytown"
        )
        if city:
            return city

    # 2️⃣ Billing Address (fallback)
    if so.billing_address:
        city = frappe.db.get_value(
            "Address",
            so.billing_address,
            "custom_citytown"
        )
        if city:
            return city

    return None


import frappe
from frappe.utils import flt


@frappe.whitelist()
def get_item_tax_template(item_code, rate, company):
    if not item_code or not company:
        return None

    rate = flt(rate)

    company_abbr = frappe.db.get_value(
        "Company",
        company,
        "abbr"
    )

    if not company_abbr:
        return None

    taxes = frappe.get_all(
        "Item Tax",
        filters={
            "parent": item_code,
            "parenttype": "Item",
            "parentfield": "taxes",
        },
        fields=[
            "item_tax_template",
            "minimum_net_rate",
            "maximum_net_rate",
        ],
        order_by="idx asc",
    )

    for tax in taxes:
        template = tax.item_tax_template

        if not template:
            continue

        min_rate = flt(tax.minimum_net_rate)
        max_rate = flt(tax.maximum_net_rate)

        rate_matches = (
            (not min_rate or rate >= min_rate)
            and
            (not max_rate or rate <= max_rate)
        )

        company_matches = template.endswith(
            f"- {company_abbr}"
        )

        if rate_matches and company_matches:
            return template

    return None


@frappe.whitelist()
def get_job_work_order_taxes_and_charges(purchase_order, company):
    if not purchase_order or not company:
        return None

    po_template = frappe.db.get_value(
        "Purchase Order",
        purchase_order,
        "taxes_and_charges"
    )

    if not po_template:
        return None

    company_abbr = frappe.db.get_value(
        "Company",
        company,
        "abbr"
    )

    if not company_abbr:
        return None

    if "RCM Out-state" in po_template:
        template = f"Output GST RCM Out-state - {company_abbr}"

    elif "RCM In-state" in po_template:
        template = f"Output GST RCM In-state - {company_abbr}"

    elif "Out-state" in po_template:
        template = f"Output GST Out-state - {company_abbr}"

    elif "In-state" in po_template:
        template = f"Output GST In-state - {company_abbr}"

    else:
        return None

    if frappe.db.exists(
        "Sales Taxes and Charges Template",
        template
    ):
        return template

    return None


def set_tax_rows_on_net_total(doc, method=None):
    for tax in doc.get("taxes") or []:
        if tax.charge_type in ("Actual", "On Previous Row Total"):
            tax.charge_type = "On Net Total"
        

